File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# class Solution:
#     def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         if not head or not head.next:
#             return head
        
#         prev, cur, temp = None, head, None

#         while cur:
#             temp = cur.next
#             cur.next = prev
#             prev = cur
#             cur = temp
        
#         return prev

# class Solution:
#     def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         return self.reverse(None, head)
    
#     def reverse(self, prev, cur):
#         if not cur:
#             return prev
        
#         nxt = cur.next
#         cur.next = prev

#         return self.reverse(cur, nxt)

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return new_head

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        ListNode dummy = new ListNode(-1);
        ListNode cur = dummy;
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>(
            (node1, node2) -> Integer.compare(node1.val, node2.val)
        );
        for (ListNode head : lists) {
            if (head!= null) {
                minHeap.offer(head);
            }
        }

        while (!minHeap.isEmpty()) {
            ListNode node = minHeap.poll();
            cur.next = node;
            if (node.next != null) {
                minHeap.offer(node.next);
            }
            cur = cur.next;
        } 
        return dummy.next;
    }
}
